File: alternative/easy/last_stone_weight.py
from typing import List
import heapq

def lastStoneWeight(stones: List[int]) -> int:
    maxHeap = []
    for stone in stones:
        heapq.heappush(maxHeap, -stone)
    while len(maxHeap) > 1:
        stone1 = -heapq.heappop(maxHeap)
        stone2 = -heapq.heappop(maxHeap)
        if stone1 != stone2:
            heapq.heappush(maxHeap, -(stone1 - stone2))
    return -maxHeap[0] if maxHeap else 0

assert lastStoneWeight([2,7,4,1,8,1]) == 1

# Algorithm - Max Heap
# Time Complexity - O(nlogn)
# Space Complexity - O(n)

# Description

# 1. Initialize a maxHeap
# 2. For each stone in stones, push the negative stone to the maxHeap
# 3. While the length of the maxHeap is greater than 1:
# 4. Pop the maxHeap and assign it to stone1
# 5. Pop the maxHeap and assign it to stone2
# 6. If stone1 is not equal to stone2, push the difference of stone1 and stone2 to the maxHeap
# 7. Return the negative of the maxHeap[0] if the maxHeap is not empty, otherwise return 0

File: src/easy/last_stone_weight.rs
#![allow(dead_code)]
use std::collections::BinaryHeap;

fn last_stone_weight(stones: Vec<i32>) -> i32 {
    let mut max_heap = BinaryHeap::from(stones);
    while max_heap.len() > 1 {
        let stone1 = max_heap.pop().unwrap();
        let stone2 = max_heap.pop().unwrap();
        if stone1 != stone2 {
            max_heap.push(stone1 - stone2);
        }
    }
    *max_heap.peek().unwrap_or(&0)
}

/*
    Algorithm - Heap

    Time    O(NlogN)
    Space   O(N)

    Description:
    1. Create a max heap
    2. Pop 2 items from the heap
    3. If they are not equal, push the difference to the heap
    4. Repeat 2-3 until the heap has 1 or 0 items
    5. Return the top item of the heap


*/

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_1() {
        let stones = vec![2, 7, 4, 1, 8, 1];
        assert_eq!(last_stone_weight(stones), 1);
    }

    #[test]
    fn test_2() {
        let stones = vec![1, 3];
        assert_eq!(last_stone_weight(stones), 2);
    }

    #[test]
    fn test_3() {
        let stones = vec![1];
        assert_eq!(last_stone_weight(stones), 1);
    }
}
